C7/MapView.cs

```diff
@@ -336,6 +322,37 @@ public class ForestLayer : LooseLayer {
 	}
 }
 
+public class MarshLayer : LooseLayer {
+	public static readonly Vector2 marshSize = new Vector2(128, 88);
+	
+	private ImageTexture largeMarshTexture;
+	private ImageTexture smallMarshTexture;
+
+	public MarshLayer() {
+		largeMarshTexture = Util.LoadTextureFromPCX("Art/Terrain/marsh.pcx", 0,   0, 512, 176);
+		smallMarshTexture = Util.LoadTextureFromPCX("Art/Terrain/marsh.pcx", 0, 176, 640, 176);
+	}
+
+	public override void drawObject(LooseView looseView, Tile tile, Vector2 tileCenter) {
+		if (tile.overlayTerrainType.name == "Marsh") {
+			int randomJungleRow = tile.yCoordinate % 2;
+			int randomMarshColumn;
+			ImageTexture marshTexture;
+			if (tile.neighborsCoast()) {
+				randomMarshColumn = tile.xCoordinate % 5;
+				marshTexture = smallMarshTexture;
+			}
+			else {
+				randomMarshColumn = tile.xCoordinate % 4;
+				marshTexture = largeMarshTexture;
+			}
+			Rect2 jungleRectangle = new Rect2(randomMarshColumn * marshSize.x, randomJungleRow * marshSize.y, marshSize);
+			Rect2 screenTarget = new Rect2(tileCenter - (float)0.5 * marshSize + new Vector2(0, -12), marshSize);
```

Comment:
Shifting the target origin by half the marsh size makes sense to me, but why does there need to an extra shift 12 pixels up? I assume there's a connection with the fact that the marsh height is 88 pixels, 24 pixels more than the usual size of 64, and 24 is double 12. So, ultimately, the delta-y is 32 pixels, same as a normal-sized texture. A comment or named variable might help make this more clear.

Honestly, this is less of me having an issue with the code than it is me being consistently annoyed with how inconsistent civ 3 terrain textures work. I think the whole system needs to reworked and consolidated into a new type of file structure.

Reply:
Yup, that's what's going on.  I moved the offset into a constant and added a comment above it to make it clearer.  It _is_ confusing.  I have a bunch of these magic offsets in my editor code as well.